### ISSUE TITLE
File usage has bad link.

### ISSUE THREAD

**Issue by jlfranklin** [backdrop/backdrop-issues#4950]
**Description of the bug**

After uploading a fie to a Backdrop site, the file shows up in admin/content/files with some links to manage it.  Clicking "Manage" gives the user three tabs, including "Manage" and "Usage".  On the "Usage" page is a link to the file itself, but the link is malformed.

**Steps To Reproduce**

To reproduce the behavior:
1. Go to admin/content/files, click "Add a file...", and add something.
2. Go back to admin/content/files and click "Manage" for the file you just uploaded.
3. Click the "Usage" tab
4. Inspect the link to the file.

**Actual behavior**

The link is to something like  "http://example.com/public://myfile.txt".  Note the "public://" in the middle of the URL.

**Expected behavior**

The link should be to "http://example.com/files/myfile.txt." 

In `file_usage_page()`, the path needs to be translated to a URL with `file_create_url()` before being passed to `l()`.

**Additional information**

Add any other information that could help, such as:
- Backdrop 1.18.1

**Comment by indigoxela** [comment]
Hi @jlfranklin, many thanks for reporting, and also providing the solution.

Confirmed on Backdrop 1.18.1 and 1.19.x-dev. The link to the file on file/XX/usage is the raw stream wrapper instead of a real url to a file.

**Comment by indigoxela** [comment]
The reason why not so many people seem to have come across this, might have to do with the fact that the problem only occurs with File entities (if the entity type using that file is "File", which is the case when uploading via the link on admin/content/files).

(And maybe the usage page (file/XX/usage) isn't the most-used in a Backdrop install :smile:)

However, it still should get fixed. A [PR](https://github.com/backdrop/backdrop/pull/3538) is available for testing and review.

**Comment by indigoxela** [comment]
Hm, thinking this over... :thinking:  The usage links for Node and Term go to the entity page (like node/[nid]), so I assume the File entity link should go to the entity page, too. A bit silly, as it's aleady linked in the primary tab right above, but still looks like more consistent.

**Comment by indigoxela** [comment]
Pinging @jlfranklin to test the PR.

This might also need a pair of eyes from the UI experts regarding where this link should actually point to. See my previous comment for explanation.

**Comment by klonos** [comment]
Not a UI expert - more like an enthusiast, but I'll chime in anyway...

First, the PR fixes the issue 👍🏼 

...second, re this:
> ...I assume the File entity link should go to the entity page, too. A bit silly, as it's aleady linked in the primary tab right above, but still looks like more consistent.

Yes, that is the most consistent and proper IMO way to be doing things 👍🏼  ...besides, the column has a title of "Entity", and if you add the same file/image to other places, like nodes/comments, you are taken to the respective entity for each entry in that usage table - you are not take to the file itself.


Last, I happened to come across these 2 related issues a couple of days ago, which I think we should look into:
- [Allow media to expose a standalone URL on a per-bundle basis](https://www.drupal.org/project/drupal/issues/3081741)
- [Media items should not be available at /media/{id} to users with 'view media' permission by default](https://www.drupal.org/project/drupal/issues/3017935)

**Comment by jlfranklin** [comment]
Every other object returns the path to its default page on the site, or `FALSE` to indicate it doesn't have a default page.  Only the `File` entity returns a URL to the object itself.  That feels like the bug is actually in the `File` entity class, not the Usage page.

Changing the `File::uri()` function to return `file/$fid` instead of `public://myfile.jpg` breaks the one test you'd expect, and seems to have no affect on uploading, managing, or using a managed file.  I was able to upload an image and use it in a Post node.  As a bonus, changing the `File::uri()` fixes this issue, too, without requiring special logic in the "Usage" page handler.

@indigoxela, what do you think?  Should we make `File` act like every other entity (at least every one in core)?

**Comment by indigoxela** [comment]
> what do you think? Should we make File act like every other entity (at least every one in core)?

If so, this should be a new issue IMO. However, I'm not sure if that's a good idea, as it would also have impact on contrib or custom implementations using that property. The fact that changing it currently only seems to breaks a single test doesn't give me enough confidence in such a big change. We could further discuss this in a new issue, though.

**Comment by klonos** [comment]
I'm tentatively marking this as RTBC, since the code adheres to coding standards, and the logic makes sense. If others review the code and are of the opinion that we need a different approach, then feel free to set the issue to "needs work".

**Comment by quicksketch** [comment]
Thanks for this conversation. @klonos taking a look at [that second link you posted](https://www.drupal.org/project/drupal/issues/3017935), Drupal actually changed the route for media entities midcycle by providing an option to switch back to the previous behavior. Though they also changed it from using `media/[id]` to the edit form. Both suggestions seem like something we might pursue for similar reasons.

I'd like to think on this one for a bit. Additional feedback is definitely welcome here.

**Comment by klonos** [comment]
> Though they also changed it from using `media/[id]` to the edit form.

Yeah, I believe that there was a security-related issue as well ...something related to media ID enumeration tactics used to "guess" URLs to media assets that should otherwise be "unpublished" (rough explanation of how I recall things).

**Comment by jlfranklin** [comment]
>  I'm not sure if that's a good idea, as it would also have impact on contrib or custom implementations using that property.

I have the same concerns, however, I have no idea if anyone even bothers with that.  Note, it's not the _property_ `$file->uri` that is going to change, it's what's returned from `File::uri()`, and barely anything seems to use that member function.

**Comment by indigoxela** [comment]
> and barely anything seems to use that member function.

That's an assumption. Decision based on assumptions can be problematic. :smirk:

**Comment by jenlampton** [comment]
> Note, it's not the property $file->uri that is going to change, it's what's returned from File::uri()

I'm not sure I like the Developer Experience of `$file->uri` having a different value than `File::uri()`...

> The usage links for Node and Term go to the entity page (like node/[nid]), so I assume the File entity link should go to the entity page, too.

At one during the addition of file_entity we had decided not to include `file/%file` in Backdrop, so that's probably how the link on the file usage page got broken. 

I prefer the PR at https://github.com/backdrop/backdrop/pull/3538, which fixes the broken link.

**Comment by quicksketch** [comment]
Really `File::uri()` and `$file->uri` won't ever match, because the method version returns an array (with keys "path" and "options"), while the property returns a string. Changing `$file->uri` would be a huge task with enormous compatibility consequences, since we inherited that from Drupal 7, I don't think that's a viable option. `File::uri()` on the other hand is much newer, less-used, and not particularly helpful at the moment, it's definitely a better candidate for changing.

In any case, I agree that's a bigger issue. I reworked the description and title of #2300 to be the canonical place for discussion on this problem, as @mikemccaffrey raised this problem quite a while ago. I put a `@todo` code comment in there linking to #2300 and merged I merged https://github.com/backdrop/backdrop/pull/3538 to solve the immediate problem. Other places that try to use `File::uri()` will still have the same problem until we fix the bigger problem of File entities being inconsistent with other entities.
